### PR TITLE
NIFI-6102 Create a set of processors to wrap and unwrap potentially malicious files so as to make them inert during transit

### PR DESCRIPTION
**Pull request by david38485**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by david38485** [comment]
https://issues.apache.org/jira/browse/NIFI-6102?filter=-2

File: nifi-nar-bundles/nifi-cloaked-dagger-bundle/nifi-cloaked-dagger-processors/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/target/
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-cloaked-dagger-bundle/nifi-cloaked-dagger-processors/.gitignore]: This should be covered in the root level gitignore.

**Comment by alopresto** [comment]
Hi David,

I have a few concerns about this PR, and hopefully you can address these points. 

I have only done a cursory review, not a line-by-line examination, but in general, I am seeing:
* A lot of duplicated code -- I think a number of the supporting classes would need to be refactored
* Some security concerns -- for example, regular Java `Random` is used when generating an encryption key rather than `SecureRandom`
* Some unit test concerns -- the `setup()` method of the tests reads in every test resource file from disk to the same fields before every test case is run

My larger concern is that this functionality is not needed within core NiFi -- it appears to be an integration with an external legacy/proprietary system. Casual internet searches for the unique names in the source code don't reveal a commonly-used toolset that NiFi would need compatibility with. Indeed, I am curious as to if this code can even be contributed as Apache License 2.0. 

From my understanding, this is basically encryption with AES/CTR/NoPadding and then optional application of a CRC32 or SHA-256 checksum. Both of these tasks (encryption & SHA-256 hashing) can be accomplished using native processors in NiFi currently. This seems like a situation that would be better handled as:

* From core NiFi -- call the external tool on the command-line using `ExecuteStreamCommand` or write a Groovy/Java script to call the APIs directly using `ExecuteScript`
* Supporting these processors -- it is perfectly acceptable to independently host these processors in a NAR bundle which individual users can download and use, but I don't think the community is prepared to support and maintain this code at this time. Many other contributors have gone this route when the use cases they had were extremely specialized

We do obviously appreciate the work that went into this contribution, but I am recommending this PR be closed.

**Comment by david38485** [comment]
> Hi David,
> 
> I have a few concerns about this PR, and hopefully you can address these points.
> 
> I have only done a cursory review, not a line-by-line examination, but in general, I am seeing:
> 
> * A lot of duplicated code -- I think a number of the supporting classes would need to be refactored
> * Some security concerns -- for example, regular Java `Random` is used when generating an encryption key rather than `SecureRandom`
> * Some unit test concerns -- the `setup()` method of the tests reads in every test resource file from disk to the same fields before every test case is run
> 
> My larger concern is that this functionality is not needed within core NiFi -- it appears to be an integration with an external legacy/proprietary system. Casual internet searches for the unique names in the source code don't reveal a commonly-used toolset that NiFi would need compatibility with. Indeed, I am curious as to if this code can even be contributed as Apache License 2.0.
> 
> From my understanding, this is basically encryption with AES/CTR/NoPadding and then optional application of a CRC32 or SHA-256 checksum. Both of these tasks (encryption & SHA-256 hashing) can be accomplished using native processors in NiFi currently. This seems like a situation that would be better handled as:
> 
> * From core NiFi -- call the external tool on the command-line using `ExecuteStreamCommand` or write a Groovy/Java script to call the APIs directly using `ExecuteScript`
> * Supporting these processors -- it is perfectly acceptable to independently host these processors in a NAR bundle which individual users can download and use, but I don't think the community is prepared to support and maintain this code at this time. Many other contributors have gone this route when the use cases they had were extremely specialized
> 
> We do obviously appreciate the work that went into this contribution, but I am recommending this PR be closed.

Andy

Thank you very much for your comments and observations. If I can answer them individually:
1) I will certainly take on board your comments on the code and tests and make sure they are incorporated.
2) This type of wrapping is something we use extensively and thought there may be a wider use case, especially between untrusted and trusted networks. 
3) I agree I had overlooked the specification publication (re your comment in the lira ticket). I  need to speak with my Technical Director as to how and where we would publish the specification for the wrapping algorithms.
4)The algorithms also encapsulate the data with reversible functions such as XOR.
5) Is there a proposal to have a 'store' for extension points that would open source, but are not provided as part of the core offering.
6) If you could provide more information (or a template) of how you could do this wrapping with standard processors I would be very interested in seeing this.

Dave

**Comment by joewitt** [comment]
Hello

It will be important to address the code level specifics of the PR to help those that might adopt it.  But I am objecting to the PR from the point of view that this is not something the community can sustain.  The specification this method appears to be about is not available for people to see/dicsuss/test the implementation against.  If that changes then it might make more sense.

In the mean time, I'd suggest this PR be closed and instead your work be made available in a Github repository your organization can offer.  We could link to it from the nifi website or wiki and there are some folks in the broader community where others can learn about it.

Please let us know if you agree and we can close this PR and JIRA.

In any event it is great that you're contributing and wanting to contribute.  While this one might not be a great candidate for inclusion hopefully there are others you're interested in that are.

Thanks

**Comment by david38485** [comment]
Joe

At this time I agree with your decision, I will correct the code as pointed out and then possibly look at a resubmission at some point in the future when the specification has been published.
Therefore please close the PR & Jira issue.

Dave

**Comment by joewitt** [comment]
thanks Dave

**Pull request closed without merging**